### PR TITLE
audio: Replace Boost with standard C++

### PR DESCRIPTION
**Pull request by argilo**
## Description
This pull requests removes all Boost from gr-audio, replacing it with standard C++:

* `boost::this_thread::sleep` -> `std::this_thread::sleep_for`
* `boost::posix_time::milliseconds` -> `std::chrono::milliseconds`
* `boost::lockfree::spsc_queue` -> `std::queue` + `std::mutex`

## Which blocks/areas does this affect?
* ALSA Audio Sink
* Windows Audio Source

## Testing Done
Unfortunately, I don't know how to build GNU Radio on Windows, so I haven't been able to test the Windows Audio Source. (The code compiles, at least. :smile:) I'd appreciate help with testing.

To test the ALSA Audio Source, I used the following flow graph:
Screenshot from 2022-03-29 22-07-40

## Checklist
- [x] I have read the [CONTRIBUTING document](https://github.com/gnuradio/gnuradio/blob/main/CONTRIBUTING.md).
- [x] I have squashed my commits to have one significant change per commit. 
- [x] I [have signed my commits before making this PR](https://github.com/gnuradio/gnuradio/blob/main/CONTRIBUTING.md#dco-signed)
- [x] My code follows the code style of this project. See [GREP1.md](https://github.com/gnuradio/greps/blob/main/grep-0001-coding-guidelines.md).
- [ ] ~~I have updated [the documentation](https://wiki.gnuradio.org/index.php/Main_Page#Documentation) where necessary.~~
- [x] ~~I have added tests to cover my changes~~, and all previous tests pass.

### CODE REVIEW

**Comment by argilo** [comment]
A couple tests I would suggest running on Windows:
* Audio Source (Sample Rate: 48k) -> Throttle (Sample Rate: 24k) -> Null Sink
  * Make sure that `buffer_queue` does not grow indefinitely.
* Audio Source (Sample Rate: 48k) -> Null Sink
  * Check whether CPU utilization is high. Right now the `work` function exits immediately if no samples are available, but I wonder whether it should instead use a condition variable to block and wait for samples to arrive through the callback function.

File: gr-audio/lib/alsa/alsa_sink.cc [hunk 1]
```diff
@@ -81,7 +83,7 @@ alsa_sink::alsa_sink(int sampling_rate, const std::string device_name, bool ok_t
         error = snd_pcm_open(&t, d_device_name.c_str(), SND_PCM_STREAM_PLAYBACK, 0);
         d_pcm_handle.set(t);
         if (error < 0) {
-            boost::this_thread::sleep(boost::posix_time::milliseconds(10));
+            std::this_thread::sleep_for(std::chrono::milliseconds(10));
```
Comment (mbr0wn) [inline gr-audio/lib/alsa/alsa_sink.cc:86]: All the cool kids do

```suggestion
            std::this_thread::sleep_for(10ms`);
```

:)

(Not a change request, just another C++ thing I kind of like)

**Comment by argilo** [comment]
I was able to test this on Windows using the conda build. Unfortunately there's a deadlock. It seems that `waveInAddBuffer`, which is called by `work` while it holds the `buffer_queue_mutex`, can call the callback function, which also needs the mutex.

I'll put this pull request in draft until that is sorted out.

**Comment by argilo** [comment]
Problem solved. The `work` function doesn't need to keep the `buffer_queue_mutex` locked once the item is popped from the `buffer_queue`, so we can unlock it at that point, and then lock it again before the loop resumes. After this change, the deadlock is gone and the Windows Audio Source appears to work correctly.

**Comment by argilo** [comment]
I also ran the two tests I suggested above (which generate excessive back-pressure or none at all) and didn't observe any problems with high CPU or memory utilization.

That said, it appears there is a bug with the Windows Audio Source, since the audio sounds bad both before and after this change. I'll look into that, but a fix probably belongs in a separate pull request.

File: gr-audio/lib/windows/windows_source.cc [hunk 2]
```diff
@@ -181,11 +183,14 @@ int windows_source::work(int noutput_items,
                 dw_items += buffer_length / 2;
             }
             buffer_queue.pop();
+            lock.unlock();
```
Comment (mormj) [inline gr-audio/lib/windows/windows_source.cc:186]: For safety, should this section be a scoped_lock?  What happens if there is an exception in those methods?

Comment (argilo) [reply gr-audio/lib/windows/windows_source.cc:186]: What safety issue did you have in mind?

My understanding is that if an exception occurs, then the `unique_lock` destructor well be called, thereby unlocking it.

**Comment by argilo** [comment]
> That said, it appears there is a bug with the Windows Audio Source, since the audio sounds bad both before and after this change. I'll look into that, but a fix probably belongs in a separate pull request.

After some further testing, I noticed the same audio issue is present even when using the PortAudio source. So I suspect this is a problem with my particular audio device, and not the Windows source block.

**Comment by willcode** [comment]
@mormj Do you think this is good for backport, based on the last couple comments here?

**Comment by mormj** [comment]
Seems like there is still a non-zero amount of risk - and the benefit to a maint release (other than limiting code divergence) would not be seen.
